### ISSUE TITLE
Fixes & refacts for v1.2.1 - [merged]

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#184]
In GitLab by @emaheux on Jan 13, 2022, 19:51

_Merges fix_zero_std_quick_and_dirty -> dev_

### What does the code in the MR do ?

Here is the changelog for this MR:
- [FIX/FEAT] Raise a LeaspyConvergenceError if variances of params collapse to zero
- [FIX] Fix some presentation issues with convergence plots
- [PERF] Slightly improve performance of Gibbs samplers
- [PERF] Initial adaptative std-dev used in Gibbs samplers is now parameter dependent (i.e. properly scaled) to speed-up convergence (cf. Work around Samplers commit)
- [FIX/FEAT] Improve and robustify `AlgorithmSettings.set_logs`
  - And improve overall code coverage by 3% by using FitOutputManager logs in various tests
- [FEAT] The computed RMSE at end of fit with Bernoulli noise model is now per feature
- [DOC] Improve documentation and implement a script to validate docstrings so to be sure they will be OK for Sphinx
- [CHORE] Clean-up code & tests
  - Fixed wrongly named helper methods `test_*` in LeaspyTestCase
  - [INTERNAL] Add some TODO for the future (esp. spot a weird behavior in mode/mean_real --> no real annealing) / remove some outdated
  - Modify a bit the tolerances used in the crossentropy (to respect torch precision for float32 and symmetry)

### Where should the reviewer start ?

May be commit per commit? (Or look at the overall changes?)
Functional tests results for fit were regenerated since it results diverge a bit from previous state (due to aforementioned PERF changes)

### How can the code be tested ?

pytest as usual or CI
You can also try to run new version on existing datasets and check for differences.
I ran on most of Winter School TPs without noticing much changes (convergence is somehow reached faster in some cases)

### When is the MR due for? (review deadline)

Let's see!

### What issues are linked to the MR ?

- On not opened issue about collapses towards zero for some *_std parameters
- On issues about documentation a bit (for docstring validator)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 13, 2022, 19:51

requested review from @little_nemo

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 13, 2022, 19:52

ping @Symcies & @PEPoulet for your information!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 13, 2022, 21:58

added 2 commits

<ul><li>6287adf6 - [TESTS] Regenerate functional tests data for calibration / mode &amp; mean_real...</li><li>670c1be4 - [RELEASE] Prepare for v1.2.1 release</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=311252046&start_sha=2d23f0069e64a763dfd1ca3bc2853148750ea7c5)

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:15

Commented on [leaspy/algo/abstract_algo.py line 45](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-23ebe00b397d9ab4e4ababa3805b3079R45)



Typo: remove the `a` after `controlled by`

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:15

Commented on [leaspy/io/data/dataframe_data_reader.py line 124](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-0ed3538185b143a62bef8f156dabe2faR124)



Typo: `The method that effectively reads [...]`

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:15

Commented on [leaspy/io/settings/algorithm_settings.py line 36](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-181e761aa9fab92bbfc7f4ac86823910R36)



Maybe don't use “personalize” since this also applies to calibration or simulation algorithms, “Depending on the algorithm you want to run / apply / setup” or something in the same spirit.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:15

Commented on [leaspy/models/univariate_model.py line 175](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-b72440adc35460de06fed3a413d0d399L175)



I looked at it, and it seems that at the moment `self.parameters['xi_mean']` is updated during the `_maximization` step of the `fit` algorithm (via the `update_model_parameters` method) to be exactly the mean of the `realizations` so in the end this shouldn't be troublesome for the general convergence of the algorithm (since both using the model parameters or computing the mean from realizations are equivalent in our iterations). I guess it prevents from computing a mean twice.

But I agree we might want to change it at some point, so that `update_MCMC_toolbox` works independently from the order of operations in the main optimization loop.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:15

Commented on [tests/functional_tests/api/test_api_fit.py line 57](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-3cac52143ebb688f46a7e2a4472983e5R57)



I might have missed part of the logic here, but are we overwriting the model if `inexistant_model` is True? If not, shouldn't we only drop the warning when `inexistant_model` is False?

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:15

Commented on [leaspy/algo/utils/samplers/algo_with_samplers.py line 42](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-3d4bf37f68727f60e5e9b50d6280c4aaR42)



Should we rather import at the beginning of the file? Moreover, unless I missed it, I don't think we actually use this package in the file.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 14, 2022, 12:25

Thank you for this MR, it is nice to have had a fix for the convergence issue that quickly! 

I made a first pass on the code, except for the first two commits (about numpydoc) for which I need a bit more time to get used to what is happening there, as well as 9e8d4e63ea4186653c00b1db496cead5d36d05bb which I still need to go through (I wasn't in a Matplotlib setup and options mood this morning..)

I will look at those three commits later on, as well as do another general pass on the MR.

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 14, 2022, 12:46

Commented on [leaspy/algo/utils/samplers/algo_with_samplers.py line 42](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-3d4bf37f68727f60e5e9b50d6280c4aaR42)



Indeed I'll fix it in next commit!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 14, 2022, 12:48

Commented on [tests/functional_tests/api/test_api_fit.py line 57](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-3cac52143ebb688f46a7e2a4472983e5R57)



I'll change the warning in case `inexistant_model` is True, but we still want a message to warn that the test is not 'complete' since we could not check consistency with expected model already saved!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 14, 2022, 12:54

Commented on [leaspy/io/settings/algorithm_settings.py line 36](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-181e761aa9fab92bbfc7f4ac86823910R36)



Indeed! Done in next commit

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 14, 2022, 12:55

Commented on [leaspy/io/data/dataframe_data_reader.py line 124](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-0ed3538185b143a62bef8f156dabe2faR124)



Thanks, done in next commit!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 14, 2022, 12:56

Commented on [leaspy/algo/abstract_algo.py line 45](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-23ebe00b397d9ab4e4ababa3805b3079R45)



Indeed, done in next commit

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 17, 2022, 14:06

Commented on [leaspy/models/univariate_model.py line 175](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-b72440adc35460de06fed3a413d0d399L175)



Then, I'm still wondering why this is not symmetrical with other parameters update in this function:
- For instance, `g` is also updated after `_maximization` step (and after burn-in: it may be different from `realizations['g'].tensor_realizations` - cf. `update_model_parameters_normal`)
  - But in turn, `update_MCMC_toolbox` is also used during `_sample_population_realizations` (for population parameters only), and I guess it'd break things to update value of `g` from `self.parameters['g']` (unlike `xi_mean`), 
  - But still, after burn-in, something seems inconsistent in this update no?

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 17, 2022, 14:15

Commented on [leaspy/models/multivariate_model.py line 267](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-44c2d4c159a4fe58d4b0ef73662defc8R267)



cf. Mattermost discussion also: @Symcies @little_nemo @PEPoulet 

More generally, I am wondering:
- when we are updating `v0` (whether here or during sampling of population parameter `v0`),
- we will then call `update_MCMC_toolbox` with param ['v0'], which will lead to re-computation of orthonormal basis, and in turn to a new mixing matrix,
- but we'll keep the previous `betas` and `sources` which were "attached" to the previous orthonormal basis and mixing matrix respectively, 
- which will ultimately modify all space-shifts of individuals in a rather uncontrolled / not isolated way,
- so shouldn't we _compensate_ the orthonormal basis change due to the change in `v0` by "re-computing" betas so the mixing matrix remains the same and so the space-shifts?

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Jan 17, 2022, 16:05

Commented on [leaspy/models/multivariate_model.py line 267](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-44c2d4c159a4fe58d4b0ef73662defc8R267)



To clarify the discussion, let's consider the MCMC-SAEM at iter k with the given three variables : A_{k}, B_{k} and C_{k}
- step 1: you update A_{k+1} given B_{k} and C_{k}
- step 2: you update B{k+1} given A_{k+1} and C_{k}
- step 3: you update C{k+1} given A_{k+1} and B_{k+1}

=> this is somehow "controlled" only by the order of the blocks in the Metropolis-Hastings within Gibbs. 


Now, what happens if we have

y = A_{k} + B_{k} + (C_{k} / B_{k})

(In our case, it is as if we could rewrite the whole `compute_tensorize_...` function as a function of `betas`, `v0` and `sources`.)

During the MCMC-SAEM, indeed, changing B_{k} would affect `y` in different manners : the second term, but also the last (=C_{k} / B_{k}). In your case, what you are proposing is to say : well, let's modifying the B_{k} but not the C_{k} / B_{k}. 

From a theoretical point of view I believe that we _don't have to_ do this, I mean there is no obligation (**). From a practical point of view, one could say that it makes the convergence go faster. 

So I think that in fact, what you are proposing is equivalent to rewrite the y above as:
 one could argue that our 

y = A_{k} + B_{k} + D_{k}

so when you change B_{k}, it does not change the last term.

Question is : are we allowed to do that when there is our orthogonality condition ?

It would need to complexify my exemples ...


PS : (**) There is no obligation to do that in the scope of the standard MCMC-SAEM but I'm not so sure as for our recentring at the end of the sampling step, where v0 <- v0 - mean(xi) ...

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 12:54

Commented on [leaspy/io/data/dataframe_data_reader.py line 124](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-0ed3538185b143a62bef8f156dabe2faR124)



changed this line in [version 3 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=313427304&start_sha=670c1be4f3c6f2701198034df415becc4b6b923a#a2576816549afb0e2a9ac85f339a13202a824506_124_124)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 12:54

Commented on [leaspy/io/settings/algorithm_settings.py line 36](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-181e761aa9fab92bbfc7f4ac86823910R36)



changed this line in [version 3 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=313427304&start_sha=670c1be4f3c6f2701198034df415becc4b6b923a#d270792e7463123a21986c7919081e50e66f17bf_36_37)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 12:54

Commented on [tests/functional_tests/api/test_api_fit.py line 57](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-3cac52143ebb688f46a7e2a4472983e5R57)



changed this line in [version 3 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=313427304&start_sha=670c1be4f3c6f2701198034df415becc4b6b923a#3cc95427ce962a533347722540e2034b019cea04_57_60)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 12:54

Commented on [leaspy/algo/utils/samplers/algo_with_samplers.py line 42](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..670c1be4f3c6f2701198034df415becc4b6b923a#diff-3d4bf37f68727f60e5e9b50d6280c4aaR42)



changed this line in [version 3 of the diff](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=313427304&start_sha=670c1be4f3c6f2701198034df415becc4b6b923a#c65eff6715a2d1b8ca19b600a5e6add2f449b60e_42_42)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 12:54

added 2 commits

<ul><li>df644e04 - WIP: Various minor fixes upon MR62 review</li><li>cf6477a5 - [TEST] Add a new test to check personalization when totally nan visits</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=313427304&start_sha=670c1be4f3c6f2701198034df415becc4b6b923a)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 12:54

marked this merge request as **draft** from df644e04195e5ca05ec2238242226e88b8175158

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 18, 2022, 13:53

added 1 commit

<ul><li>9c794a8f - [FEAT] Add a flag to discard empty column warning in DataFrameReader</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=313475676&start_sha=cf6477a51d40313b039e86c0d8fdc5e7fc9f9d70)

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 19, 2022, 16:25

Commented on [leaspy/io/logs/visualization/plotter.py line 297](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-c9c3894d7799e493bdc4bbb8b3bdefb2R297)



This is minor and maybe not relevant for the scope of this MR because the class lacks documentation in general, but this function is poorly “self-documented”, since `max_patient_number` can in fact be either an `int` or a list of indices of patients. Maybe, waiting for a more extensive documentation of this class, we could change the argument name to `patients_used` which would be consistent with both possible use-case of the method.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 19, 2022, 16:32

I reviewed the 3 commits that I had not reviewed yet. 

Note that for the first commit (ae81df1ec7e8e40b4de7eb0f97bc655af8b24017), I was able to understand the global logic of the new docstring tests -- which appears both to be and run fine -- but I have not been able to follow through all the implementation details yet. I can allocate more time to review this first commit if you think it is necessary.

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 19, 2022, 16:33

Commented on [leaspy/io/logs/visualization/plotter.py line 297](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-c9c3894d7799e493bdc4bbb8b3bdefb2R297)



Yes I totally agree, `Plotter` really lacks a proper documentation... but also a complete refactorization (cf. #51 & #45). I think it's not really worth it documentating before starting the refactorization even if it's a bit a pity :/

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 19, 2022, 17:37

Thank you!

For the docstrings validator script, I think it'll be OK then (I also tested this script successfully in other project and most of the code is inspired from sklearn own script to do the same thing!)

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 24, 2022, 11:14

Commented on [leaspy/models/univariate_model.py line 175](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-b72440adc35460de06fed3a413d0d399L175)



I took another look at it and I now agree that there might be some inconsistency during the post burn-in phase, since `self.parameters['xi_mean']` embeds some “memory” of the previous iterations while `realizations['g']` does not. 

I tried to unroll the consequences of changing the update to `values['xi_mean'] = torch.mean(realizations['xi'].tensor_realization)`, since I was unsure as to why the update was performed using the `self.parameters['xi_mean']` in the first place, and I don't really see any issue and it would make more sense to me. I don't know how much it would affect calibration results though.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Jan 24, 2022, 11:22

Alright, therefore regarding the fixes and workflows already introduced by this MR, this is good to me, I can approve the current changes. :slight_smile: 

Yet, should we wait until we reach some consensus on the two remaining threads, or move them to new issues? They might involve a bit more “in-depth” discussion and go out of the MR's scope. What do you think?

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 26, 2022, 17:41

Commented on [leaspy/models/multivariate_model.py line 267](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-44c2d4c159a4fe58d4b0ef73662defc8R267)



Sorry for replying so late:

Indeed I found your demonstration convincing: with this way of presenting things I must admit the current behavior seems totally legit! I think that what puzzled me is the fact that, unlike all the other population parameters, here `v0` is deeply entangled with individual parameters, namely `sources` (through the orthonormal_basis / mixing_matrix).
As you also mention, trying to "uncouple" v0 from the remaining parameters _may_ speed-up a bit the convergence, but still it is only based on intuition and we do not have any guarantee.

> Question is : are we allowed to do that when there is our orthogonality condition ?

In fact, I thought again at it and my proposal ("shouldn't we *compensate* the orthonormal basis change due to the change in `v0` by _re-computing_ betas so the mixing matrix remains the same and so the space-shifts?") was indeed not correct! as soon as the new `v0` is not collinear to the previous `v0`, their orthogonal spaces will be different, and thus there may be cases where we could not find any betas that would span the **same** mixing matrix as before the `v0` update (and in particular if `Ns = N - 1`, we will never be able to do it).

> PS : (\*\*) There is no obligation to do that in the scope of the standard MCMC-SAEM but I'm not so sure as for our recentring at the end of the sampling step, where v0 <- v0 - mean(xi) ...

Indeed, in fact this last point also seems tricky to me (in fact this was the very root of my doubts!). It seems to me that your previous arguments are not legit any more after the end of the sampling, and that the current operation is not correct... I'm a bit puzzled here on what we should do?!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 26, 2022, 17:59

Commented on [leaspy/models/univariate_model.py line 175](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-b72440adc35460de06fed3a413d0d399L175)



I also took another look at it and as far as I understand:
- for individual parameters (such as `xi`):
  - the current behaviour seems legit to me (updating from `self.parameters`), because those were already updated just before, whether directly from realizations during the "burn-in phase" or post "burn-in")
  - (updating directly from realizations post burn-in seems incorrect to me for xi_mean, since we would "de-synchronize" the MCMC version of the model from the "true" model)
- for population parameters (such as `g`):
  - the current behavior (updating directly from realizations) seems legit to me **only during the memory-less part of the algorithm**
  - post burn-in (part that embeds some memory), we would have to "alternately":
    - update directly from realizations, otherwise `GibbsSampler._sample_population_realizations` would not be correct
    - update from `self.parameters` after the `_maximization_step` otherwise there would be a gap between the self.parameters and the "MCMC version of the parameters"

Anyway I think this topic should deserve a dedicated issue & a dedicated fix!
What do others think about this @Symcies @PEPoulet ?

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 26, 2022, 18:12

Sorry I was a bit long to answer the two remaining threads! I also agree that we'd better split the two remaining threads into distinct issues as their comprehension & (proper) resolution need a bit more of discussion (and won't be that easy I guess). 

Let's wait a couple more days to converge on diagnoses concerning those issues and create the needed issues!

**Comment by NicolasGensollen** [comment]
In GitLab by @PEPoulet on Jan 27, 2022, 12:35

Commented on [leaspy/models/multivariate_model.py line 267](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-44c2d4c159a4fe58d4b0ef73662defc8R267)



I'm really asking myself similar questions in a very different context (coding MALA and Riemannian MALA samplers), in the sense that I don't particularly benefit from the sequential sampling of the Gibbs method. The idea of Gibbs normally applies well in practice when the block of varaibles are "independent". 

-> In our case the model needs first to stabilize around the best v0 before having a nice estimation of the betas. 

The question that comes to mind to solve this : why not sample v0 and betas at the same time ? Well for a general use case this won't work because the proposal will almost never be accepted. BUT with drift based samplers, the proposal is pushed towards the gradient of the likelihood so it is more likely to be accepted. More on this topic in the weeks to come...

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jan 31, 2022, 18:29

added 1 commit

<ul><li>6465cb6c - Micro fix in Plotter.plot_patient_reconstructions when less patients in dataset than requested</li></ul>

[Compare with previous version](/icm-institute/aramislab/leaspy/-/merge_requests/62/diffs?diff_id=321851421&start_sha=34eefc6329d91f0cce9b9179d08920e9352ce042)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 1, 2022, 10:38

@little_nemo @Symcies @PEPoulet what about creating 2 issues for the 2 remaining threads that needs a bit more of discussion/R&D, merging this MR and releasing the 1.2.1 (still waiting for critical fix on AlgorithmsSettings.save not working - which is critical for leaspype `calibrate_resampling` for instance)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 1, 2022, 10:39

marked this merge request as **ready**

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Feb 1, 2022, 15:56

I agree. I will just review the recent commits that you added and merging afterward is okay for me!

I can open the corresponding remaining issues if you want.

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Feb 1, 2022, 16:24

approved this merge request

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Feb 1, 2022, 16:42

Commented on [leaspy/models/multivariate_model.py line 267](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-44c2d4c159a4fe58d4b0ef73662defc8R267)



created #59 to continue this discussion

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on Feb 1, 2022, 16:53

Commented on [leaspy/models/univariate_model.py line 175](https://github.com/aramis-lab/leaspy/compare/8d69c2b1b7cf4f92ae6ce213cc4aadd41214a118..6465cb6c9777702f3654eb8dae8182e6a0c9f204#diff-b72440adc35460de06fed3a413d0d399L175)



created #60 to continue this discussion

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 2, 2022, 19:27

resolved all threads

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 3, 2022, 12:09

@Symcies can you approve this MR in addition to Nemo please ? (please don't pay attention to commits from !63 that were already approved and merged in this branch!)

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Feb 3, 2022, 12:13

approved this merge request

**Comment by NicolasGensollen** [comment]
In GitLab by @Symcies on Feb 3, 2022, 12:14

Sorry Gitlab didn't noticed me as I was not in the reviewers. Done now :)

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 3, 2022, 15:45

Thanks!

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 3, 2022, 15:45

resolved all threads

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Feb 3, 2022, 15:45

mentioned in commit 87470db1115ef983250985ce3d419b66cfec9d06
